File: Deepspace-Ruby/lib/Dice.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

class Dice
    
  
  def initialize
    @NHANGARSPROB=0.25
    @NSHIELDSPROB=0.25
    @NWEAPONSPROB=0.33
    @FIRSTSHOTPROB=0.5
    @EXTRAEFFICIENCYPROB=0.8
    @generator=Random.new
  end
  
  def extraEfficiency
    p=@generator.rand(0...1.0)
    if p< @EXTRAEFFICIENCYPROB
      true
    else
      false
    end
  end
  
  def initWithNHangars
    resultado=@generator.rand(0.0...1.0)
    if resultado<@NHANGARSPROB
      0
    else 
      1
    end
  end
  
  def initWithNWeapons
    resultado=@generator.rand(0.0...1.0)
    if resultado<@NWEAPONSPROB
      0
    elsif resultado<2*@NWEAPONSPROB
        1
    else 2
    end
  end
  
  def initWithNShields
    resultado=@generator.rand(0.0...1.0)
    if resultado <@NSHIELDSPROB
      0
    else 1
    end
  end
  
  def whoStarts(nPlayers)
    v=@generator.rand(0...nPlayers)
    v
  end
  
  def firstShot 
    resultado=@generator.rand(0.0...1.0)
    if resultado< @FIRSTSHOTPROB
      GameCharacter::SPACESTATION
    else GameCharacter::ENEMYSTARSHIP
    end
  end
  
  def spaceStationMoves(speed)
    resultado=@generator.rand(0.0...1.0)
    if resultado< speed
      true
    else false
    end
  end
  
  def to_s
    return  "NHANGARSPROB: #{@NHANGARSPROB} NSHIELDSPROB: #{@NSHIELDSPROB}, NWEAPONSPROB: #{@NWEAPONSPROB},
      FIRSTSHOTPROB: #{@FIRSTSHOTPROB}"
  end
end

File: Deepspace-Java/src/deepspace/Dice.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.Random;

/**
 *
 * @author inaki
 */
public class Dice {
  private final float NHANGARSPROB;
  private final float NSHIELDSPROB;
  private final float NWEAPONSPROB;
  private final float FIRSTSHOTPROB;
  private final float EXTRAEFFICIENCYPROB;
  private Random generator;
    
  Dice (){
      NHANGARSPROB=0.25f;
      NSHIELDSPROB=0.25f;
      NWEAPONSPROB=0.33f;
      FIRSTSHOTPROB=0.5f;
      EXTRAEFFICIENCYPROB=0.8f;
      generator=new Random();      
  }
  
  public int initWithNHangars (){
      float r=generator.nextFloat();
      if (r <=NHANGARSPROB)
          return 0;
      else return 1;
  }
  
  public int initWithNWeapons (){
      float r=generator.nextFloat();
      if (r<=NWEAPONSPROB)
          return 1;
      else if (r<=2*NWEAPONSPROB)
          return 2;
      else return 3;
  }
  
  public int initWhithNShields (){
      float r=generator.nextFloat();
      if (r<=NSHIELDSPROB)
          return 0;
      else return 1;
  }
  
  public int whoStarts(int nPlayers){
      int r=generator.nextInt(nPlayers);
      return r;
  }
  
  public GameCharacter firstShot(){
      float r=generator.nextFloat();
      if ( r <= FIRSTSHOTPROB)
        return GameCharacter.SPACESTATION ;
      else return GameCharacter.ENEMYSTARSHIP;
  }
  
  public boolean spaceStationMoves(float speed){
      float r=generator.nextFloat();
      
      return(r < speed);
 
  }
  
  public boolean extraEfficiency(){
      float r=generator.nextFloat();
      return r<=EXTRAEFFICIENCYPROB;
  }

    @Override
    public String toString() {
        return "Dice{" + "generator=" + generator + '}';
    }
  

  
  
  
  
}
